# OptimiR
miRSeq data alignment workflow - integrates genetic information to assess the impact of variants on miRNA expression

## Requirements

The following python libraries must be installed: pysam and biopython (available through pip).
The following programs must be installed: bowtie2, cutadapt, samtools. If the path of these programs is not in $PATH, you can provide it directly to OPTIMIR (see ./OPTIMIR --help).
As of nov. 2018 OptimiR has been only used (successfuly) in a GNU/Linux environment. Works fine with python2 and python3.

## Usage

./OPTIMIR --fq /path/to/sample.fq(.gz) --vcf /path/to/genotypes.vcf --dir_output /path/to/output/dir/

Different options are available to customize the reference library and the outputs. Use ./OPTIMIR -h to see all options. The user can provide a vcf file (and is recommanded to do so) with a list of variants, and OptimiR will extract variants located on miRNAs (a.k.a polymiRs) and generate new alternative sequences of these polymiRs integrating the alternative version of the variant.

In dir_output, the following directories are created:
 - OptimiR_lib : contains the alignment index generated by OptimiR, as well as a pickle directory containing python object (which allow the library to not be regenerated at each run if the inputed files did not change), and a fasta file which contains all the miRs and polymiRs sequences generated by OptimiR
 - OptimiR_tmp : contains temporary files generated during OptimiR run. The final alignment file is available in 3_PostProcess in bam format. If --rmTempFiles option is given to OPTIMIR, these directories will be deleted. If OptimiR is run on the same sample more than once and OptimiR_tmp not deleted, the sample is not trimmed again, which saves computation time.
 - OptimiR_Results : contains the abundances tables for each miRs / polymiRs, and detailed counts of isomiRs. If genotype is provided, a consistency table is also generated and indicate the consistency between the alignment on a polymiR and the genotypes of the variants involved. Annotation files are also generated to provide information on cross-mapping reads that could not be resolved by OptimiR, and another give an indication on the parental hairpins from which the reads are originating (computed from templated tailed nucleotides).
  
## Summary Files

If several samples must be processed, it is recommanded to provide the same output directory for all runs. The python script OPTIMIR_SUMMARY with the output directory as a parameter to produce summary files for all runs (abundances tables and annotation tables).

## Run on an example

2 fastq samples and a vcf file are provided with OptimiR, in directory "example". 2 scripts allow the user to run OptimiR on these examples, either on a single sample (example_1.sh) or on all samples (example_all.sh). Just launch the script as follow:
./example_1.sh or ./example_all.sh

## About the .vcf file

The vcf file is optional. All variants provided in the vcf file that are mapping on miRNAs will be integrated in new sequences generated by OptimiR. Reads will be allowed to align to these new alternative mIRNA sequences.
The genotype can also be provided by the user in the vcf file, in which case OptimiR will perform a genotype consistency step and retain only alignments on polymIRs that are consistent with the genotype.
The sample name in the vcf file must match the name of the provided fastq (minus the extension of the fastq file)
When a vcf with genotypes is provided, 2 tables : consistency_table and polymiRs_table are generated in the outputs with summary alignments on each polymiR.

### A few limitations:
 - Multi-allelic variants must be provided on two separate lines. Allele string must not contain a comma, and genotype format must have format 0/0, 0/1, 1/0 or 1/1 but no 0/2 and such... (btw: also works with phased variants format).
 - Combination of indels, or indels and SNPs: if a SNP and an indel map on the same miRNA, the combination of both my not be generated, in which case a warning will be displayed during library preparation. Combinations of SNPs works fine.

## OPTIMIR -h

usage: OPTIMIR [-h] -i FASTQ [-o OUTDIR] [-g VCF] [--seedLen SEEDLEN]
               [--w5 WEIGHT5] [--scoreThresh SCORE_THRESH]
               [--consistentRate INCONSISTENT_THRESHOLD] [--rmTempFiles]
               [--annot ANNOT_FILES] [--adapt3 ADAPT3] [--adapt5 ADAPT5]
               [--readMin READMIN] [--readMax READMAX] [--bqThresh BQTHRESH]
               [--trimAgain] [--maturesFasta MATURES]
               [--hairpinsFasta HAIRPINS] [--gff3 GFF3] [--quiet]
               [--cutadapt CUTADAPT] [--bowtie2 BOWTIE2]
               [--bowtie2_build BOWTIE2_BUILD] [--samtools SAMTOOLS]

OptimiR: A bioinformatics pipeline designed to detect and quantify miRNAs,
isomiRs and polymiRs from miRSeq data, & study the impact of genetic
variations on polymiRs' expression

optional arguments:
  -h, --help            show this help message and exit
  -i FASTQ, --fq FASTQ  Full path of the sample fastq file (accepted formats
                        and extensions: fastq, fq and fq.gz)
  -o OUTDIR, --dirOutput OUTDIR
                        Full path of the directory where output files are
                        generated
  -g VCF, --vcf VCF     Full path of the vcf file with genotypes
  --seedLen SEEDLEN     Choose the alignment seed length used in option '-L'
                        by Bowtie2 (default:17)
  --w5 WEIGHT5          Choose the weight applied on events detected on the 5'
                        end of aligned reads (default:4)
  --scoreThresh SCORE_THRESH
                        Choose the threshold for alignment score above which
                        alignments are discarded (default:9)
  --consistentRate INCONSISTENT_THRESHOLD
                        Choose the rate threshold for inconsistent reads
                        mapped to a polymiR above which the alignment is
                        flagged as highly suspicious (default:0.01)
  --rmTempFiles         Add this option to remove temporary files (trimmed
                        fastq, collapsed fastq, mapped reads, annotated bams
  --annot ANNOT_FILES   Control which annotation file is produced by adding
                        corresponding letter : 'h' for expressed_hairpins, 'p'
                        for polymiRs_table, 'i' for consistency_table, 'c' for
                        remaining_ambiguous, 's' for isomiRs_dist. Ex: '--
                        annot hpics' (default) will produce all of them
  --adapt3 ADAPT3       Define the 3' adaptor sequence (default is NEB &
                        ILLUMINA: AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC -a
                        TGGAATTCTCGGGTGCCAAGG -> hack: use -a to add adapter
                        sequences
  --adapt5 ADAPT5       Define the 5' adaptor sequence (default is NEB:
                        ATCTACACGTTCAGAGTTCTACAGTCCGACGATC
  --readMin READMIN     Define the minimum read length defined with option -m
                        in cutadapt (default: 15)
  --readMax READMAX     Define the maximum read length defined with option -M
                        in cutadapt (default: 27)
  --bqThresh BQTHRESH   Define the Base Quality threshold defined with option
                        -q in cutadapt (default: 28)
  --trimAgain           Add this option to trim files that have been trimmed
                        in a previous application. By default, when temporary
                        files are kept, trimmed files are reused. If you wish
                        to change a paramater used in the trimming step of the
                        workflow, this parameter is a must.
  --maturesFasta MATURES
                        Path to the reference library containing mature miRNAs
                        sequences (default: from miRBase v21)
  --hairpinsFasta HAIRPINS
                        Path to the reference library containing pri-miRNAs
                        sequences (default: from miRBase v21)
  --gff3 GFF3           Path to the reference library containing miRNAs and
                        pri-miRNAs coordinates (default: from miRBase v21,
                        GRCh38 coordinates)
  --quiet               Add this option to remove OptimiR progression on
                        screen
  --cutadapt CUTADAPT   Provide path to the cutadapt binary
  --bowtie2 BOWTIE2     Provide path to the bowtie2 binary
  --bowtie2_build BOWTIE2_BUILD
                        Provide path to the bowtie2 index builder binary
  --samtools SAMTOOLS   Provide path to the samtools binary


## Troubleshooting

